WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,147 @@
+import Foundation
+import Yosemite
+import class WooFoundation.CurrencyFormatter
+
+enum CardPresentPaymentResult {
+    case success(Order)
+    case failure(CardPaymentErrorProtocol)
+    case cancellation
+}
+
+class CardPresentPaymentsAdaptor {
+    private let currencyFormatter: CurrencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)
+    private let siteID: Int64
+
+    // These connection controllers shouldn't be lazy, I'm just being lazy.
+    // They're lazy because of `self` implementing `alertsPresenter`.
+    // If that were done by another owned struct, and the handler passed to it as and when it changed, we could do this in the init.
+    private lazy var bluetoothConnectionController: CardReaderConnectionController = {
+        CardReaderConnectionController(
+            forSiteID: siteID,
+            knownReaderProvider: CardReaderSettingsKnownReaderStorage(),
+            alertsPresenter: self,
+            alertsProvider: BluetoothReaderConnectionAlertsProvider(),
+            configuration: CardPresentConfigurationLoader().configuration,
+            analyticsTracker: CardReaderConnectionAnalyticsTracker(
+                configuration: CardPresentConfigurationLoader().configuration,
+                siteID: siteID,
+                connectionType: .userInitiated, // This will need a tweak – it changes from use to use.
+                analytics: ServiceLocator.analytics))
+    }()
+
+    private lazy var tapToPayConnectionController: BuiltInCardReaderConnectionController = {
+        BuiltInCardReaderConnectionController(
+            forSiteID: siteID,
+            alertsPresenter: self,
+            alertsProvider: BluetoothReaderConnectionAlertsProvider(),
+            configuration: CardPresentConfigurationLoader().configuration,
+            analyticsTracker: CardReaderConnectionAnalyticsTracker(
+                configuration: CardPresentConfigurationLoader().configuration,
+                siteID: siteID,
+                connectionType: .userInitiated, // This will need a tweak – it changes from use to use.
+                analytics: ServiceLocator.analytics))
+    }()
+
+    var paymentsAlertHandler: CardPresentPaymentsAlertHandling?
+
+    init(siteID: Int64) {
+        self.siteID = siteID
+    }
+
+    func collectPayment(for order: Order,
+                        using discoveryMethod: CardReaderDiscoveryMethod) async -> CardPresentPaymentResult {
```

Comment:
Would it make sense to switch over a result type here? 
```suggestion
                        using discoveryMethod: CardReaderDiscoveryMethod) async -> Result<Order, CardPresentPaymentError> {
```
The enum can be simplified and moved away as well: 

```diff
- enum CardPresentPaymentResult {
-    case success(Order)
+ enum CardPresentPaymentError: Error {
    case failure(CardPaymentErrorProtocol)
    case cancellation
}

```

And the continuation would also be clearer:
``` diff
- continuation.resume(returning: .cancellation)
+ continuation.resume(returning: .failure(.cancellation))
```

Reply:
We could, and I probably wouldn't even pick it up at review if someone did that... but personally I don't really like to use `Result` for tri-states like this. 

Cancellation isn't an error or failure, it's a user action and we're doing what they want us to. I think that making it a failure tends to confuse the call site, as you have to unwrap the errors and check whether they were actually cancellations before you handle them.